timeseries/src/autogluon/timeseries/models/chronos/model.py

```diff
@@ -0,0 +1,366 @@
+import logging
+import os
+from itertools import chain
+from pathlib import Path
+from typing import Any, Dict, Literal, Optional, Union
+
+import numpy as np
+import pandas as pd
+from transformers import AutoConfig, AutoModelForCausalLM, AutoModelForSeq2SeqLM
+
+from autogluon.common.loaders import load_pkl
+from autogluon.timeseries.dataset.ts_dataframe import ITEMID, TIMESTAMP, TimeSeriesDataFrame
+from autogluon.timeseries.models.abstract import AbstractTimeSeriesModel
+from autogluon.timeseries.utils.forecast import get_forecast_horizon_index_ts_dataframe
+
+from .chronos import ChronosConfig, ChronosPipeline, ChronosPretrainedModel
+
+# NOTE: We avoid imports for torch and lightning.pytorch at the top level and hide them inside class methods.
+# This is done to skip these imports during multiprocessing (which may cause bugs)
+
+logger = logging.getLogger(__name__)
+
+
+# allowed HuggingFace model paths with custom parameter definitions
+MODEL_CONFIGS = {
+    "amazon/chronos-t5-tiny": {
+        "num_gpus": 0,  # minimum number of required GPUs
+    },
+    "amazon/chronos-t5-mini": {"num_gpus": 0},
+    "amazon/chronos-t5-small": {"num_gpus": 1},
+    "amazon/chronos-t5-base": {"num_gpus": 1},
+    "amazon/chronos-t5-large": {"num_gpus": 1},
+}
+
+
+class ChronosInferenceDataset:
+    """A container for time series datasets that implements the ``torch.utils.data.Dataset`` interface"""
+
+    def __init__(
+        self,
+        target_df: TimeSeriesDataFrame,
+        context_length: int,
+        target_column: str = "target",
+    ):
+        assert target_df is not None
+        assert target_df.freq, "Initializing GluonTS data sets without freq is not allowed"
+
+        self.target_column = target_column
+        self.context_length = context_length
```

Comment:
Nit: A potential optimization to save GPU memory is to set `context_length = min(context_length, min(indices_sizes))` (i.e., no longer than the shortest TS in the dataset)

Reply:
Do you mean longest TS in the dataset? 

So in the current construction, by time the data is provided the model will already be initialized. i.e., as above, we're not doing deferred initialization for the model.

Reply:
You are right, sorry, I meant the longest TS in the dataset.

As far as I understood, we are instantiating a new `ChronosInferenceDataset` every time `model.predict(data)` is called. I meant `context_length` as a property of this `ChronosInferenceDataset` object that affects how much padding we add (or how short we cut all the time series). If user calls `predict` with `data` where all TS have length 100, the current code will pad them all to 512, which would increase the memory consumption. If we set `context_length = min(context_length, max(indices_sizes))`, then `context_length` will be automatically shortened to 100.

This might be a premature optimization that is not needed, so feel free to ignore my comment.

Reply:
Included this logic. Looks a bit shaky to me in hindsight, it might use some @shchur reviewing 👀 .